service/collector/endpoint.go

```diff
@@ -117,17 +116,23 @@ func (e *Endpoint) Collect(ch chan<- prometheus.Metric) error {
 		if !isNginxIngressControllerInstalled(appList) {
 			// When nginx ingress controller is not deployed, worker ports are not open so there is no need to check for connectivity
 			e.logger.LogCtx(ctx, "level", "debug", "message", fmt.Sprintf("skipping cluster %#q since it doesn't have %#q installed", kvmConfig.Spec.Cluster.ID, nginxIngressControllerAppName))
+
 			continue
 		}
 
-		g.Go(func() error {
-			endpoint, err := e.k8sClient.CoreV1().Endpoints(clusterID).Get(ctx, workerEndpoint, getOpts)
-			if err != nil {
-				return microerror.Mask(err)
-			}
+		endpoint, err := e.k8sClient.CoreV1().Endpoints(clusterID).Get(ctx, workerEndpoint, getOpts)
```

Comment:
Currently it's running the whole logic in one goroutine which makes it super easy to go over the scraping limit. Current prometheus scrape limit is 10s, we had two endpoints that were unresponsive (took us 5s each to timeout) and prometheus was reporting ingress-exporter as down.

Reply:
It looks like it's currently running one goroutine for each cluster. Running one per worker endpoint IP instead makes sense to me. My only concern would be spawning a large number of goroutines that all need to use the network. What's the most concurrent requests we could have (i.e. total number of worker endpoint IPs across all clusters)? Should we use a channel to limit it to some maximum number? I guess if it's under 100 or so it shouldn't be a problem.

Reply:
Good point, limited the concurrent goroutines to 30 to play it safe